Copyright (c) Microsoft Corporation. All rights reserved. 
Licensed under the MIT License.

![Impressions](https://PixelServer20190423114238.azurewebsites.net/api/impressions/MachineLearningNotebooks/how-to-use-azureml/machine-learning-pipelines/parallel-run/tabular-dataset-partition-per-column.png)

# Using Azure Machine Learning Pipelines for Batch Inference for tabular input partitioned by column value

In this notebook, we will demonstrate how to make predictions on large quantities of data asynchronously using the ML pipelines with Azure Machine Learning. Batch inference (or batch scoring) provides cost-effective inference, with unparalleled throughput for asynchronous applications. Batch prediction pipelines can scale to perform inference on terabytes of production data. Batch prediction is optimized for high throughput, fire-and-forget predictions for a large collection of data.

> **Tip**
If your system requires low-latency processing (to process a single document or small set of documents quickly), use [real-time scoring](https://docs.microsoft.com/en-us/azure/machine-learning/service/how-to-consume-web-service) instead of batch prediction.

This example will create a partitioned tabular dataset by splitting the rows in a large csv file by its value on specified column. Each partition will form up a mini-batch in the parallel processing procedure.

The outline of this notebook is as follows:

- Create a tabular dataset partitioned by value on specified column.
- Do batch inference on the dataset with each mini-batch corresponds to one partition.

## Prerequisites
If you are using an Azure Machine Learning Notebook VM, you are all set. Otherwise, make sure you go through the configuration Notebook located at https://github.com/Azure/MachineLearningNotebooks first. This sets you up with a working config file that has information on your workspace, subscription id, etc. 

### Connect to workspace

In [1]:
from azureml.core.workspace import Workspace

from azureml.core.authentication import InteractiveLoginAuthentication
from azureml.core import Workspace

subscription_id = '0e9bace8-7a81-4922-83b5-d995ff706507'
# Azure Machine Learning resource group NOT the managed resource group
resource_group = 'azureml' 

#Azure Machine Learning workspace name, NOT Azure Databricks workspace
workspace_name = 'ws01ent'  
tenant_id ='72f988bf-86f1-41af-91ab-2d7cd011db47' 

auth = InteractiveLoginAuthentication(tenant_id =tenant_id)
# Instantiate Azure Machine Learning workspace
ws = Workspace.get(name=workspace_name,
                   subscription_id=subscription_id,
                   resource_group=resource_group,auth= auth)

print('Workspace name: ' + ws.name, 
      'Azure region: ' + ws.location, 
      'Subscription id: ' + ws.subscription_id, 
      'Resource group: ' + ws.resource_group, sep = '\n')

datastore = ws.get_default_datastore()

C:\Users\janguy\Anaconda3\envs\dlresearch\lib\site-packages\numpy\_distributor_init.py:30: UserWarning: loaded more than 1 DLL from .libs:
C:\Users\janguy\Anaconda3\envs\dlresearch\lib\site-packages\numpy\.libs\libopenblas.NOIJJG62EMASZI6NYURL6JBKM4EVBGM7.gfortran-win_amd64.dll
C:\Users\janguy\Anaconda3\envs\dlresearch\lib\site-packages\numpy\.libs\libopenblas.PYQHXLVVQ7VESDPUVUADXEVJOBGHJPAY.gfortran-win_amd64.dll
  warnings.warn("loaded more than 1 DLL from .libs:\n%s" %
WARNING - Warning: Falling back to use azure cli login credentials.
If you run your code in unattended mode, i.e., where you can't give a user input, then we recommend to use ServicePrincipalAuthentication or MsiAuthentication.
Please refer to aka.ms/aml-notebook-auth for different authentication mechanisms in azureml-sdk.


Workspace name: ws01ent
Azure region: westus2
Subscription id: 0e9bace8-7a81-4922-83b5-d995ff706507
Resource group: azureml


In [2]:
import azureml.core
print(azureml.core.VERSION)

1.31.0


### Download OJ sales data from opendataset url

In [ ]:
# ! pip install azureml-opendatasets --user

In [3]:
import os
from azureml.opendatasets import OjSalesSimulated
dataset_maxfiles = 10 # Set to 11973 or 0 to get all the files

# Pull all of the data
oj_sales_files = OjSalesSimulated.get_file_dataset()

# Pull only the first `dataset_maxfiles` files
if dataset_maxfiles:
    oj_sales_files = oj_sales_files.take(dataset_maxfiles)

# Create a folder to download
download_path = '..\..\data\oj_sales_data' 
os.makedirs(download_path, exist_ok=True)

# Download the data
oj_sales_files.download(download_path, overwrite=True,)

['c:\\Users\\janguy\\OneDrive - Microsoft\\Documents\\projects\\training\\Many-Models\\data\\oj_sales_data\\https%3A\\%2Fazureopendatastorage.azurefd.net\\ojsales-simulatedcontainer\\oj_sales_data\\Store1000_dominicks.csv',
 'c:\\Users\\janguy\\OneDrive - Microsoft\\Documents\\projects\\training\\Many-Models\\data\\oj_sales_data\\https%3A\\%2Fazureopendatastorage.azurefd.net\\ojsales-simulatedcontainer\\oj_sales_data\\Store1000_minute.maid.csv',
 'c:\\Users\\janguy\\OneDrive - Microsoft\\Documents\\projects\\training\\Many-Models\\data\\oj_sales_data\\https%3A\\%2Fazureopendatastorage.azurefd.net\\ojsales-simulatedcontainer\\oj_sales_data\\Store1000_tropicana.csv',
 'c:\\Users\\janguy\\OneDrive - Microsoft\\Documents\\projects\\training\\Many-Models\\data\\oj_sales_data\\https%3A\\%2Fazureopendatastorage.azurefd.net\\ojsales-simulatedcontainer\\oj_sales_data\\Store1001_dominicks.csv',
 'c:\\Users\\janguy\\OneDrive - Microsoft\\Documents\\projects\\training\\Many-Models\\data\\oj_sales_

### Upload OJ sales data to datastore

In [4]:
target_path = 'oj_sales_data'

datastore.upload(src_dir = download_path,
                target_path = target_path,
                overwrite = True, 
                show_progress = True)


Uploading an estimated of 10 files
Uploading ..\..\data\oj_sales_data\https%3A\%2Fazureopendatastorage.azurefd.net\ojsales-simulatedcontainer\oj_sales_data\Store1000_minute.maid.csv
Uploaded ..\..\data\oj_sales_data\https%3A\%2Fazureopendatastorage.azurefd.net\ojsales-simulatedcontainer\oj_sales_data\Store1000_minute.maid.csv, 1 files out of an estimated total of 10
Uploading ..\..\data\oj_sales_data\https%3A\%2Fazureopendatastorage.azurefd.net\ojsales-simulatedcontainer\oj_sales_data\Store1000_tropicana.csv
Uploaded ..\..\data\oj_sales_data\https%3A\%2Fazureopendatastorage.azurefd.net\ojsales-simulatedcontainer\oj_sales_data\Store1000_tropicana.csv, 2 files out of an estimated total of 10
Uploading ..\..\data\oj_sales_data\https%3A\%2Fazureopendatastorage.azurefd.net\ojsales-simulatedcontainer\oj_sales_data\Store1001_dominicks.csv
Uploaded ..\..\data\oj_sales_data\https%3A\%2Fazureopendatastorage.azurefd.net\ojsales-simulatedcontainer\oj_sales_data\Store1001_dominicks.csv, 3 files out

$AZUREML_DATAREFERENCE_4eb0d09ea48548ea8c04115672931893

### Create tabular dataset
Create normal tabular dataset

In [5]:
from azureml.core import Dataset

dataset = Dataset.Tabular.from_delimited_files(path=(datastore, 'oj_sales_data/*/*/*/*/*.csv'))
print(dataset.to_pandas_dataframe())

     WeekStarting  Store      Brand  Quantity  Advert  Price   Revenue
0      1990-06-14   1000  dominicks     12003       1   2.59  31087.77
1      1990-06-21   1000  dominicks     10239       1   2.39  24471.21
2      1990-06-28   1000  dominicks     17917       1   2.48  44434.16
3      1990-07-05   1000  dominicks     14218       1   2.33  33127.94
4      1990-07-12   1000  dominicks     15925       1   2.01  32009.25
...           ...    ...        ...       ...     ...    ...       ...
1205   1992-09-03   1003  dominicks     10302       1   1.94  19985.88
1206   1992-09-10   1003  dominicks     13502       1   2.16  29164.32
1207   1992-09-17   1003  dominicks     19644       1   2.67  52449.48
1208   1992-09-24   1003  dominicks     13860       1   2.29  31739.40
1209   1992-10-01   1003  dominicks     11040       1   1.99  21969.60

[1210 rows x 7 columns]


### Partition the tabular dataset
Partition the dataset by column 'store' and 'brand'. You can get a partition of data by specifying the value of one or more partition keys. E.g., by specifying `store=1000 and brand='tropicana'`, you can get all the rows that matches this condition in the dataset.

In [6]:
partitioned_dataset = dataset.partition_by(partition_keys=['Store', 'Brand'], target=(datastore, "partition_by_key_res"), name="partitioned_oj_data")
partitioned_dataset.partition_keys

WARNING - Method partition_by: This is an experimental method, and may change at any time. Please see https://aka.ms/azuremlexperimental for more information.


Validating arguments.
Arguments validated.
Uploading file to /partition_by_key_res/de780196-0cb4-4a48-bfcd-20a1a3039fa2/
Successfully uploaded file to datastore.
Creating a new dataset.
Successfully created a new dataset.
registering a new dataset.
Successfully created and registered a new dataset.


['Store', 'Brand']

### Create or Attach existing compute resource

In [7]:
import os
from azureml.core.compute import AmlCompute, ComputeTarget

# choose a name for your cluster
compute_name = os.environ.get("AML_COMPUTE_CLUSTER_NAME", "cpu-cluster")
compute_min_nodes = os.environ.get("AML_COMPUTE_CLUSTER_MIN_NODES", 0)
compute_max_nodes = os.environ.get("AML_COMPUTE_CLUSTER_MAX_NODES", 2)

# This example uses CPU VM. For using GPU VM, set SKU to STANDARD_NC6
vm_size = os.environ.get("AML_COMPUTE_CLUSTER_SKU", "STANDARD_D2_V2")


if compute_name in ws.compute_targets:
    compute_target = ws.compute_targets[compute_name]
    if compute_target and type(compute_target) is AmlCompute:
        print('found compute target. just use it. ' + compute_name)
else:
    print('creating a new compute target...')
    provisioning_config = AmlCompute.provisioning_configuration(vm_size = vm_size,
                                                                min_nodes = compute_min_nodes, 
                                                                max_nodes = compute_max_nodes)

    # create the cluster
    compute_target = ComputeTarget.create(ws, compute_name, provisioning_config)
    
    # can poll for a minimum number of nodes and for a specific timeout. 
    # if no min node count is provided it will use the scale settings for the cluster
    compute_target.wait_for_completion(show_output=True, min_node_count=None, timeout_in_minutes=20)
    
     # For a more detailed view of current AmlCompute status, use get_status()
    print(compute_target.get_status().serialize())

WARNING - Class SynapseCompute: This is an experimental class, and may change at any time. Please see https://aka.ms/azuremlexperimental for more information.
WARNING - Class SynapseCompute: This is an experimental class, and may change at any time. Please see https://aka.ms/azuremlexperimental for more information.


found compute target. just use it. cpu-cluster


### Intermediate/Output Data

In [8]:
from azureml.pipeline.core import Pipeline, PipelineData

output_dir = PipelineData(name="inferences", datastore=datastore)

### Calculate total revenue of each mini-batch partitioned by dataset partition key(s)
The script sum up the total revenue of a mini-batch.

In [9]:
scripts_folder = "..\..\code"
inference_script_file = "aml_prs\prediction.py"
train_script_file = "aml_prs\model_train.py"

# peek at contents
with open(os.path.join(scripts_folder, train_script_file)) as train_file:
    print(train_file.read())

# Copyright (c) Microsoft. All rights reserved.
# Licensed under the MIT license.

import os
from azureml.core import Run, Model
import joblib
from util.timeseries_utilities import ColumnDropper, SimpleLagger, SimpleCalendarFeaturizer, SimpleForecaster
from sklearn.metrics import mean_squared_error, mean_absolute_error
from sklearn.linear_model import LinearRegression
import numpy as np
import pandas as pd
def init():
    global ws
    current_run = Run.get_context()
    ws = current_run.experiment.workspace

    print("Init complete")


def run(mini_batch):
    print(f'run method start: {__file__}, run({mini_batch})')
    target_column= 'Quantity'
    timestamp_column= 'WeekStarting'
    drop_columns=['Revenue', 'Store', 'Brand']
    #Get the store and brand. They are unique from the group so just the first value is sufficient
    store = str(mini_batch['Store'].iloc[0])
    brand = str(mini_batch['Brand'].iloc[0])

    model_name="prs_"+store+"_"+brand
    test_size=20
    # 1.0 Form

## Build and run the batch inference pipeline
### Specify the environment to run the script
You would need to specify the required private azureml packages in dependencies. 

In [10]:
from azureml.core import Environment
from azureml.core.runconfig import CondaDependencies, DEFAULT_CPU_IMAGE

batch_conda_deps = CondaDependencies.create(pip_packages=['sklearn', 'pandas', 'joblib', 'azureml-defaults', 'azureml-core', 'azureml-dataprep[fuse]'])
batch_env = Environment(name="many_models_environment")
batch_env.python.conda_dependencies = batch_conda_deps
batch_env.docker.base_image = DEFAULT_CPU_IMAGE

In [11]:
from azureml.pipeline.steps import ParallelRunStep, ParallelRunConfig

# In a real-world scenario, you'll want to shape your process per node and nodes to fit your problem domain.
parallel_run_train_config = ParallelRunConfig(
    source_directory=scripts_folder,
    entry_script=train_script_file,  # the user script to run against each input
    partition_keys=['Store', 'Brand'],
    error_threshold=-1,
    output_action='append_row',
    append_row_file_name="training_output.txt",
    environment=batch_env,
    compute_target=compute_target, 
    node_count=2,
    run_invocation_timeout=600
)

WARNING - Parameter partition_keys: This is an experimental parameter, and may change at any time. Please see https://aka.ms/azuremlexperimental for more information.


In [12]:
parallel_run_train_step = ParallelRunStep(
    name='train',
    inputs=[partitioned_dataset.as_named_input("partitioned_tabular_input")],
    output=output_dir,
    parallel_run_config=parallel_run_train_config,
    allow_reuse=False
)

C:\Users\janguy\Anaconda3\envs\dlresearch\lib\site-packages\azureml\pipeline\core\_parallel_run_step_base.py:591: UserWarning: 
ParallelRunStep requires azureml-dataset-runtime[fuse,pandas] for tabular dataset.
Please add relevant package in CondaDependencies.
  warnings.warn(


In [13]:
from azureml.core import Experiment
from azureml.pipeline.core import Pipeline

pipeline = Pipeline(workspace=ws, steps=[parallel_run_train_step])

pipeline_run = Experiment(ws, 'Many_Model_Forecast').submit(pipeline)

Created step train [7ec698e8][82a5e60e-3634-4395-9af4-26bd6cbfff70], (This step will run and generate new outputs)
Submitted PipelineRun 3a46572c-94e0-4f06-9e90-91a664b4f71f
Link to Azure Machine Learning Portal: https://ml.azure.com/runs/3a46572c-94e0-4f06-9e90-91a664b4f71f?wsid=/subscriptions/0e9bace8-7a81-4922-83b5-d995ff706507/resourcegroups/azureml/workspaces/ws01ent&tid=72f988bf-86f1-41af-91ab-2d7cd011db47


In [14]:
pipeline_run.wait_for_completion(show_output=True)

PipelineRunId: 3a46572c-94e0-4f06-9e90-91a664b4f71f
Link to Azure Machine Learning Portal: https://ml.azure.com/runs/3a46572c-94e0-4f06-9e90-91a664b4f71f?wsid=/subscriptions/0e9bace8-7a81-4922-83b5-d995ff706507/resourcegroups/azureml/workspaces/ws01ent&tid=72f988bf-86f1-41af-91ab-2d7cd011db47
PipelineRun Status: Running


StepRunId: fa553678-a2e7-480d-aafc-660714e7b37e
Link to Azure Machine Learning Portal: https://ml.azure.com/runs/fa553678-a2e7-480d-aafc-660714e7b37e?wsid=/subscriptions/0e9bace8-7a81-4922-83b5-d995ff706507/resourcegroups/azureml/workspaces/ws01ent&tid=72f988bf-86f1-41af-91ab-2d7cd011db47
StepRun( train ) Status: Running

Streaming azureml-logs/55_azureml-execution-tvmps_9dad2b059b79040f8ad5ca8131190ad001a3d68c0a40c6795d0e83cf5f33f2c0_d.txt
2021-06-23T19:49:27Z Running following command: /bin/bash -c sudo blobfuse /mnt/batch/tasks/shared/LS_root/jobs/ws01ent/azureml/fa553678-a2e7-480d-aafc-660714e7b37e/mounts/workspaceblobstore --tmp-path=/mnt/batch/tasks/shared/LS_ro

'Finished'

In [18]:
import pandas as pd
import tempfile

batch_run = pipeline_run.find_step_run(parallel_run_train_step.name)[0]
batch_output = batch_run.get_output_data(output_dir.name)

target_dir = tempfile.mkdtemp()
batch_output.download(local_path=target_dir)
result_file = os.path.join(target_dir, batch_output.path_on_datastore, parallel_run_train_config.append_row_file_name)

df = pd.read_csv(result_file, delimiter=" ", header=None)

df.columns = [ "Store", "Brand", "mse", "mape", "rmse", "model_name"]
print("Train result has ", df.shape[0], " rows")
df.head(10)

Train result has  10  rows


Store        Brand           mse       mape         rmse  \
0   1000  minute.maid  1.022011e+07  18.172361  3196.891234   
1   1001    tropicana  8.552595e+06  19.463062  2924.481928   
2   1000    dominicks  1.080181e+07  23.172431  3286.611380   
3   1001    dominicks  8.920132e+06  20.350161  2986.658951   
4   1001  minute.maid  1.556051e+07  26.075884  3944.680670   
5   1003    dominicks  1.278600e+07  23.512086  3575.751535   
6   1002    tropicana  1.069882e+07  20.099309  3270.904584   
7   1000    tropicana  9.844161e+06  19.846187  3137.540593   
8   1002    dominicks  9.404042e+06  23.352023  3066.601004   
9   1002  minute.maid  6.130244e+06  17.805611  2475.932918   

             model_name  
0  prs_1000_minute.maid  
1    prs_1001_tropicana  
2    prs_1000_dominicks  
3    prs_1001_dominicks  
4  prs_1001_minute.maid  
5    prs_1003_dominicks  
6    prs_1002_tropicana  
7    prs_1000_tropicana  
8    prs_1002_dominicks  
9  prs_1002_minute.maid

### Create the configuration to wrap the inference script
The parameter `partition_keys` is a list containing a subset of the dataset partition keys, specifying how is the input dataset partitioned. Each and every possible combination of values of partition_keys will form up a mini-batch. E.g., by specifying `partition_keys=['store', 'brand']` will result in mini-batches like `store=1000 && brand=tropicana`, `store=1000 && brand=dominicks`, `store=1001 && brand=dominicks`, ...

Run(Experiment: Many_Model_Forecast,
Id: 72bb2e27-20f0-42bd-b0eb-e0ef7b57020b,
Type: azureml.StepRun,
Status: Completed)

In [21]:
from azureml.pipeline.steps import ParallelRunStep, ParallelRunConfig

# In a real-world scenario, you'll want to shape your process per node and nodes to fit your problem domain.
parallel_run_inference_config = ParallelRunConfig(
    source_directory=scripts_folder,
    entry_script=inference_script_file,  # the user script to run against each input
    partition_keys=['Store', 'Brand'],
    error_threshold= -1,
    output_action='append_row',
    append_row_file_name="prediction_output.txt",
    environment=batch_env,
    compute_target=compute_target, 
    node_count=2,
    run_invocation_timeout=600
)

WARNING - Parameter partition_keys: This is an experimental parameter, and may change at any time. Please see https://aka.ms/azuremlexperimental for more information.


### Create the pipeline step

In [22]:
parallel_run_inference_step = ParallelRunStep(
    name='forecast',
    inputs=[partitioned_dataset.as_named_input("partitioned_tabular_input")],
    output=output_dir,
    parallel_run_config=parallel_run_inference_config,
    allow_reuse=False
)

### Run the pipeline

In [23]:
from azureml.core import Experiment
from azureml.pipeline.core import Pipeline

pipeline = Pipeline(workspace=ws, steps=[parallel_run_inference_step])

pipeline_run = Experiment(ws, 'Many_Model_Forecast').submit(pipeline)

Created step forecast [f86ab7e3][7dd799f0-a27b-4420-ae09-c4f66c28fe4d], (This step will run and generate new outputs)
Submitted PipelineRun dac5b96c-b823-4293-8729-4d146cc0feec
Link to Azure Machine Learning Portal: https://ml.azure.com/runs/dac5b96c-b823-4293-8729-4d146cc0feec?wsid=/subscriptions/0e9bace8-7a81-4922-83b5-d995ff706507/resourcegroups/azureml/workspaces/ws01ent&tid=72f988bf-86f1-41af-91ab-2d7cd011db47


In [24]:
pipeline_run.wait_for_completion(show_output=True)

PipelineRunId: dac5b96c-b823-4293-8729-4d146cc0feec
Link to Azure Machine Learning Portal: https://ml.azure.com/runs/dac5b96c-b823-4293-8729-4d146cc0feec?wsid=/subscriptions/0e9bace8-7a81-4922-83b5-d995ff706507/resourcegroups/azureml/workspaces/ws01ent&tid=72f988bf-86f1-41af-91ab-2d7cd011db47
PipelineRun Status: Running


StepRunId: d07f1ba7-5d15-47ab-85da-61a8853a986a
Link to Azure Machine Learning Portal: https://ml.azure.com/runs/d07f1ba7-5d15-47ab-85da-61a8853a986a?wsid=/subscriptions/0e9bace8-7a81-4922-83b5-d995ff706507/resourcegroups/azureml/workspaces/ws01ent&tid=72f988bf-86f1-41af-91ab-2d7cd011db47
StepRun( forecast ) Status: NotStarted
StepRun( forecast ) Status: Running

Streaming azureml-logs/55_azureml-execution-tvmps_9dad2b059b79040f8ad5ca8131190ad001a3d68c0a40c6795d0e83cf5f33f2c0_d.txt
2021-06-23T19:54:59Z Running following command: /bin/bash -c sudo blobfuse /mnt/batch/tasks/shared/LS_root/jobs/ws01ent/azureml/d07f1ba7-5d15-47ab-85da-61a8853a986a/mounts/workspaceblobstor

'Finished'

## View the prediction results
In the total_income.py file above you can see that the ResultList with the filename and the prediction result gets returned. These are written to the DataStore specified in the PipelineData object as the output data, which in this case is called inferences. This containers the outputs from all of the worker nodes used in the compute cluster. You can download this data to view the results ... below just filters to the first 10 rows

In [26]:
import pandas as pd
import tempfile

batch_run = pipeline_run.find_step_run(parallel_run_inference_step.name)[0]
batch_output = batch_run.get_output_data(output_dir.name)

target_dir = tempfile.mkdtemp()
batch_output.download(local_path=target_dir)
result_file = os.path.join(target_dir, batch_output.path_on_datastore, parallel_run_config.append_row_file_name)

df = pd.read_csv(result_file, delimiter=" ", header=None)
df.columns = ["WeekStarting", "Prediction", "Store", "Brand"]
print("Prediction has ", df.shape[0], " rows")
df.head(10)

Prediction has  1210  rows


WeekStarting    Prediction  Store      Brand
0   1990-06-14           NaN   1001  dominicks
1   1990-06-21           NaN   1001  dominicks
2   1990-06-28           NaN   1001  dominicks
3   1990-07-05           NaN   1001  dominicks
4   1990-07-12  13554.387879   1001  dominicks
5   1990-07-19  15375.931186   1001  dominicks
6   1990-07-26  15243.409687   1001  dominicks
7   1990-08-02  13649.643619   1001  dominicks
8   1990-08-09  14017.612778   1001  dominicks
9   1990-08-16  15976.467391   1001  dominicks